frontend/packages/hardhat/contracts/EthernalChatIncentivized.sol

```diff
@@ -5,95 +5,136 @@ import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+		bytes32 newMerkleRoot,
+		bytes32 merkleRootOfAppendedData
+	) public {
+		require(cid != bytes32(0), "Invalid CID");
+		DataInfo memory dataInfo = mapDataInfo[msg.sender];
+		require(
+			dataInfo.cid == bytes32(0) ||
+				dataInfo.merkleRoot == merkleRootOfAppendedData,
+			"Data not appended correctly"
+		);
+		require(newMerkleRoot != bytes32(0), "invalid merkle root");
+
+		dataInfo.cid = cid;
+		dataInfo.numberOfChunks = numberOfChunks;
+		dataInfo.sizeOfChunks = sizeOfChunks;
+		dataInfo.merkleRoot = newMerkleRoot;
+
+		emit CIDUpdated(msg.sender, cid);
+	}
 
-    // function getChallenge(address addr) public view returns ( uint64 index ) OnlyStorageProvider { // return index
-    // // require ..
-    //     index = 0 ; // random;
-    // }
+	/// @notice Add funds to an existing DataInfo by sending tokens and updating
+	/// @param newPrice The new price in tokens
+	/// @dev Make sure the DataInfo is non zero (has been already created)
+	function addFundsForStorage(uint256 newPrice) public {
+		DataInfo memory dataInfo = mapDataInfo[msg.sender];
+		require(dataInfo.cid != bytes32(0), "DataInfo not found");
+		paymentToken.transferFrom(msg.sender, address(this), newPrice);
+		dataInfo.allocatedTokens += newPrice;
+	}
 
+	/// @notice Retrieves the CID (Content Identifier) stored for the user.
+	/// @return The CID associated with the user.
+	/// @dev Passes when the user has a CID stored. This function is view only, so it doesn't modify the state of the contract
+	function getCID() public view returns (bytes32) {
+		DataInfo memory dataInfo = mapDataInfo[msg.sender];
+		require(dataInfo.cid != bytes32(0), "No CID found for this address");
+		return dataInfo.cid;
+	}
 
-    /// @notice Allows storage to get a partial amount of reward in token, if it hasn't been already taken in this amount of time 
-    function getStorageReward() public view returns (bytes32) {
-        // bytes32 cid = addressToCID[msg.sender];
-        // require( cid.length > 0, "No CID found for this address");
-        // // Check if the chunk size given correspond to the expected size
-        // return cid;
-    }
+	function getChallenge(
+		address addr
+	) public view OnlyStorageProvider(addr) returns (uint64 index) {
+		require(
+			mapDataInfo[addr].cid != bytes32(0),
+			"No DataInfo found for this address"
+		);
+		index = uint64(
+			uint256(
+				keccak256(abi.encodePacked(block.timestamp, block.prevrandao))
+			) % mapDataInfo[addr].numberOfChunks
+		);
+	}
 
-    /// @notice Take out all the rewards in Eth based on the amount of tokens the address holds
-    function withdrawRewards() public view returns (bytes32) {
-        // bytes32 cid = addressToCID[msg.sender];
-        // require( cid.length > 0, "No CID found for this address");
-        // return cid;
-    }
+	/// @notice Allows storage to get a partial amount of reward in token, if it hasn't been already taken in this amount of time
+	function getStorageReward() public view returns (uint256) {
```

Comment:
That is my bad I didn't explained, I just copy pasted it, I didn't had the time to define the functions properly (only the name is good haha): this is not a view function, we need to transfer some tokens or modify some state. My bet was to modify a state variable "tokensEarned" ... but I realize the state change has no interest over giving directly eth in term of gas cost..

The reason I think it is better to have two functions: getStorageReward (which will update the state) and then withdrawRewards is because it is a better incentive: we can only withdraw after the specified amount of time (let say a month) and yo have to accumulate: this give you a stake system where the StorageProvider will have money at stake and an incentive to continue. 

So I think we can have a state variable that will keep the data information that will move the allocated tokens to the tokens earned by the storageProvider

Reply:
We also need to create a verify proof function and make sure the provider gets his proofs verified before getting his tokens... This means we will need to add arguments

Reply:
This should be name "verifyChallenge" actually